firedrake/mesh.py

```diff
@@ -816,11 +891,52 @@ def make_offset(self, entity_dofs, ndofs):
                     dof_offset[i] = entity_offset[b]
         return dof_offset
 
-    @property
+    @utils.cached_property
     def layers(self):
         """Return the number of layers of the extruded mesh
         represented by the number of occurences of the base mesh."""
-        return self._layers
+        if self.variable_layers:
+            raise ValueError("Can't ask for mesh layers with variable layers")
+        else:
+            return self.cell_set.layers
+
+    @utils.cached_property
+    def variable_layers(self):
+        """Does the extruded mesh have variable layer numbers?"""
+        return not self.cell_set.constant_layers
+
+    def entity_layers(self, height, label=None):
+        """Return the number of layers on each entity of a given plex
+        height.
+
+        :arg height: The height of the entity to compute the number of
+           layers (0 -> cells, 1 -> facets, etc...)
+        :arg label: An optional label name used to select points of
+           the given height (if None, then all points are used).
+        :returns: a numpy array of the number of layers on the asked
+           for entities (or a single layer number for the constant
+           layer case).
+        """
+        if self.variable_layers:
+            return extnum.entity_layers(self, height, label)
+        else:
+            return self.cell_set.layers
+
+    @utils.cached_property
+    def layer_extents(self):
+        """The layer extents for all mesh points.
+
+        For variable layers, the layer extent does not match those for cells.
+
+        :returns: numpy array of layer extents (in PyOP2 format
+            :math:`[start, stop)`), of shape ``(num_cells, 4)`` where
+            the first two extents are used for allocation and the last
+            two for iteration.
+        """
+        extents = extnum.layer_extents(self)
```

Comment:
It does not help to make this a `cached_property` since if it can be provided, then it is always needed.